docs/developer-guide/deprecation.md

```diff
@@ -0,0 +1,265 @@
+
+```python
+from zope.deprecation import deprecated
+
+foo = None
+foo = deprecated(foo, "foo is no more, use bar instead")
+```
+
+
+### Move functions and classes
+
+This example describes how to move some classes or functions from a Python file at {file}`old/foo/bar.py` to {file}`new/baz/baaz.py`.
+Here, `zope.deferredimport` offers a deprecation helper.
+It also avoids circular imports at initialization time.
+
+```python
+import zope.deferredimport
+zope.deferredimport.initialize()
+
+zope.deferredimport.deprecated(
+    "Import from new.baz.baaz instead",
+    SomeOldClass="new.baz:baaz.SomeMovedClass",
+    some_old_function="new.baz:baaz.some_moved_function",
+)
+
+def some_function_which_is_not_touched_at_all():
+    pass
+```
+
+### Deprecate a GenericSetup profile
+
+In GenericSetup, the `post_handler` attribute in ZCML can be used to call a function after the profile was applied.
+Use this feature to issue a warning.
+
+First, register the same profile twice, under both the new name and old.
+
+```xml
+<genericsetup:registerProfile
+    name="default"
+    title="My Fancy Package"
+    directory="profiles/default"
+    description="..."
+    provides="Products.GenericSetup.interfaces.EXTENSION"
+    />
+
+<genericsetup:registerProfile
+    name="some_confusing_name"
+    title="My Fancy Package (deprecated)"
+    directory="profiles/some_confusing_name"
+    description="... (use profile default instead)"
+    provides="Products.GenericSetup.interfaces.EXTENSION"
+    post_handler=".setuphandlers.deprecate_profile_some_confusing_name"
+    />
+```
+
+Then in {file}`setuphandlers.py`, add a function.
+
+```python
+import warnings
+
+def deprecate_profile_some_confusing_name(tool):
+    warnings.warn(
+        'The profile with id "some_confusing_name" was renamed to "default".',
+        DeprecationWarning
+    )
+```
+
+### Deprecate a template position
+
+Sometimes you need to move templates to new locations.
+Since add-ons often use [`z3c.jbot`](https://github.com/zopefoundation/z3c.jbot) to override templates by their position, you'll need to point them to the new position as well as make sure that the override still works with the old position.
+
+To deprecate a package, follow these steps.
+
+1.  In the old package folder's {file}`__init__.py`, add a dictionary `jbot_deprecations` that maps the old template locations to their new counterparts.
+
+    ```python
+    jbot_deprecations = {
+        "plone.locking.browser.info.pt": "plone.app.layout.viewlets.info.pt"
+    }
+    ```
+
+1.  Add this deprecation snippet to the package {file}`configure.zcml` file.
+
+    ```{code-block} xml
+    :emphasize-lines: 7-14
+    :linenos:
+    
+    <configure
+        xmlns="http://namespaces.zope.org/zope"
+        xmlns:browser="http://namespaces.zope.org/browser"
+        xmlns:zcml="http://namespaces.zope.org/zcml"
+        >
+    
+      <include
+          zcml:condition="installed z3c.jbot"
+          package="z3c.jbot"
```

Comment:
We should explicitly load `meta.zcml` of z3c.jbot`:

```suggestion
      <include
          package="z3c.jbot"
          file="meta.zcml"
          zcml:condition="installed z3c.jbot"
```

That is the file that defines the `jbot-deprecations` feature and the `jbotDeprecated` directive.
This is how [`plone.locking` does it](https://github.com/plone/plone.locking/blob/4.0.0a3/src/plone/locking/configure.zcml#L14-L23).